lib/services/useracl.go

```diff
@@ -0,0 +1,189 @@
+	// DB defines access to database resource.
+	DB ResourceAccess `json:"db"`
+	// KubeServers defines access to kubernetes servers.
+	KubeServers ResourceAccess `json:"kubeServers"`
+	// Desktops defines access to desktops.
+	Desktops ResourceAccess `json:"desktops"`
+	// AccessRequests defines access to access requests.
+	AccessRequests ResourceAccess `json:"accessRequests"`
+	// Billing defines access to billing information.
+	Billing ResourceAccess `json:"billing"`
+	// ConnectionDiagnostic defines access to connection diagnostics.
+	ConnectionDiagnostic ResourceAccess `json:"connectionDiagnostic"`
+	// Clipboard defines whether the user can use a shared clipboard during windows desktop sessions.
+	Clipboard bool `json:"clipboard"`
+	// DesktopSessionRecording defines whether the user's desktop sessions are being recorded.
+	DesktopSessionRecording bool `json:"desktopSessionRecording"`
+	// DirectorySharing defines whether a user is permitted to share a directory during windows desktop sessions.
+	DirectorySharing bool `json:"directorySharing"`
+	// Download defines whether the user has access to download Teleport Enterprise Binaries
+	Download ResourceAccess `json:"download"`
+	// Download defines whether the user has access to download the license
+	License ResourceAccess `json:"license"`
+	// Plugins defines whether the user has access to manage hosted plugin instances
+	Plugins ResourceAccess `json:"plugins"`
+	// Integrations defines whether the user has access to manage integrations.
+	Integrations ResourceAccess `json:"integrations"`
+	// DeviceTrust defines access to device trust.
+	DeviceTrust ResourceAccess `json:"deviceTrust"`
+	// Locks defines access to locking resources.
+	Locks ResourceAccess `json:"lock"`
+	// Assist defines access to assist feature.
+	Assist ResourceAccess `json:"assist"`
+	// SAMLIdpServiceProvider defines access to `saml_idp_service_provider` objects.
+	SAMLIdpServiceProvider ResourceAccess `json:"samlIdpServiceProvider"`
+}
+
+func hasAccess(roleSet RoleSet, ctx *Context, kind string, verbs ...string) bool {
+	for _, verb := range verbs {
+		// Since this check occurs often and does not imply the caller is trying to
+		// ResourceAccess any resource, silence any logging done on the proxy.
+		if err := roleSet.GuessIfAccessIsPossible(ctx, apidefaults.Namespace, kind, verb, true); err != nil {
+			return false
+		}
+	}
+	return true
+}
+
+func newAccess(roleSet RoleSet, ctx *Context, kind string) ResourceAccess {
+	return ResourceAccess{
+		List:   hasAccess(roleSet, ctx, kind, types.VerbList),
+		Read:   hasAccess(roleSet, ctx, kind, types.VerbRead),
+		Edit:   hasAccess(roleSet, ctx, kind, types.VerbUpdate),
+		Create: hasAccess(roleSet, ctx, kind, types.VerbCreate),
+		Delete: hasAccess(roleSet, ctx, kind, types.VerbDelete),
+		Use:    hasAccess(roleSet, ctx, kind, types.VerbUse),
+	}
+}
+
+func NewUserACL(user types.User, userRoles RoleSet, features proto.Features, desktopRecordingEnabled bool) UserACL {
```

Comment:
> It seemed to me that extracting the whole `NewUserACL` function is better than extracting only `hasAccess` and `newAccess` - we can use a more abstract function.

Yeah, I think it makes sense, given that `hasAccess` and `newAccess` have only been called from within the old equivalent of `NewUserACL` and we want to share the logic in `NewUserACL` between lib/web and lib/teleterm.